q2_mlab/doctor.py

```diff
@@ -0,0 +1,220 @@
+
+
+def get_inserted_results(results_dir):
+    inserted_dir = os.path.join(results_dir, "inserted")
+    if not os.path.isdir(inserted_dir):
+        return []
+    else:
+        return [
+            f.name for f in os.scandir(inserted_dir) if f.path.endswith(".qza")
+        ]
+
+
+def filter_duplicate_parameter_results(
+    list_of_artifact_paths, results_dir, delete=False
+):
+    artifact_paths = list_of_artifact_paths.copy()
+
+    # Change working dirs
+    prev_working_dir = os.getcwd()
+    os.chdir(results_dir)
+
+    if len(artifact_paths) == 0:
+        raise ValueError("There are no result artifacts to remove.")
+    artifact_paths.sort()
+
+    curr_path = artifact_paths[0]
+    curr_param_idx = int(curr_path.split("_")[0])
+
+    for next_path in artifact_paths[1:]:
+        next_param_idx = int(next_path.split("_")[0])
+        if next_param_idx == curr_param_idx:
+            # Remove the older of the duplicate results
+            curr_path_time = datetime.fromtimestamp(
+                os.stat(curr_path).st_mtime
+            )
+            next_path_time = datetime.fromtimestamp(
+                os.stat(next_path).st_mtime
+            )
+            if curr_path_time < next_path_time:
+                artifact_paths.remove(curr_path)
+                if delete:
+                    os.remove(curr_path)
+            else:
+                artifact_paths.remove(next_path)
+                if delete:
+                    os.remove(next_path)
+                continue  # Do not update current_path
+        # Update current path
+        curr_path = next_path
+        curr_param_idx = next_param_idx
+    
+    # Change working dir back
+    os.chdir(prev_working_dir)
+
+    return artifact_paths
+
+
+def doctor_hyperparameter_search(
+    dataset,
+    preparation,
+    target,
+    algorithm,
+    base_dir,
+    max_results=1000,
+    delete_duplicates=False,
+):
+    """
+    Searches for the job description file for the experiment described by
+    this function's parameters, searches through the files produced for
+    missing results, and relaunches jobs necessary for completing those results.
+
+            Parameters:
+                    dataset (str): Name of dataset
+                    preparation (str): Name of data type/preparation (e.g. 16S)
+                    target (str): Name of the target variable in the metadata
+                    algorithm (str): Valid algorithm included in q2_mlab
+                    base_dir (str): The directory in which create the file structure.
+                    max_results (str): The maximum number of result artifacts to expect.
+                    delete_duplicates (bool): If True, deletes the older of duplicated results on the same parameter id.
+            Returns:
+                    summary (str): Summary of errors and jobs relaunched/to be relaunched
+    """
+    barnacle_out_dir = os.path.join(base_dir, dataset, "barnacle_output/")
```

Comment:
this seems specific to our system? Maybe restructure/rename so that the code here is not explicitly dependent on our system

Reply:
I removed this since it is never actually used here, but `barnacle_out_dir` is baked into other parts of the codebase so I'll need to change that eventually.